File: genium/src/test/java/com/here/genium/generator/cbridge/CBridgeTypeMapperTest.java
/*
 * Copyright (C) 2016-2018 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cbridge;

import static com.here.genium.generator.cbridge.CBridgeNameRules.BASE_HANDLE_IMPL_FILE;
import static com.here.genium.generator.cbridge.CBridgeNameRules.BASE_REF_NAME;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

import com.here.genium.common.FrancaTypeHelper;
import com.here.genium.generator.cpp.CppLibraryIncludes;
import com.here.genium.generator.cpp.CppNameResolver;
import com.here.genium.model.cbridge.CBridgeIncludeResolver;
import com.here.genium.model.cbridge.CType;
import com.here.genium.model.common.Include;
import com.here.genium.model.cpp.CppIncludeResolver;
import com.here.genium.model.franca.DefinedBy;
import org.franca.core.franca.*;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Answers;
import org.mockito.Mock;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

@RunWith(PowerMockRunner.class)
@PrepareForTest({FrancaTypeHelper.class, DefinedBy.class})
public final class CBridgeTypeMapperTest {

  private static final Include CPP_INCLUDE = Include.Companion.createInternalInclude("CPP_INCLUDE");
  private static final Include CBRIDGE_INCLUDE =
      Include.Companion.createInternalInclude("CBRIDGE_INCLUDE");

  @Mock private FTypeRef francaTypeRef;
  @Mock private FTypeRef francaTypeRef2;
  @Mock private FTypeDef francaTypeDef;
  @Mock private FStructType francaStructType;
  @Mock private FTypeCollection francaTypeCollection;
  @Mock private FEnumerationType francaEnum;

  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
  private FMapType francaMap;

  @Mock private CppIncludeResolver cppIncludeResolver;
  @Mock private CBridgeIncludeResolver includeResolver;
  @Mock private CppNameResolver cppNameResolver;

  private CBridgeTypeMapper typeMapper;

  @Before
  public void setUp() {
    mockStatic(FrancaTypeHelper.class, DefinedBy.class);

    typeMapper =
        new CBridgeTypeMapper(
            cppIncludeResolver, cppNameResolver, includeResolver, "::FooHash", null);

    when(francaStructType.getName()).thenReturn("someStruct");

    when(DefinedBy.findDefiningTypeCollection(any(FModelElement.class)))
        .thenReturn(francaTypeCollection);

    when(cppIncludeResolver.resolveInclude(any())).thenReturn(CPP_INCLUDE);
    when(includeResolver.resolveInclude(any())).thenReturn(CBRIDGE_INCLUDE);
  }

  @Test
  public void mapTypeDefStruct() {
    when(francaTypeRef.getDerived()).thenReturn(francaTypeDef);
    when(francaTypeDef.getActualType()).thenReturn(francaTypeRef2);
    when(francaTypeRef2.getDerived()).thenReturn(francaStructType);
    when(cppNameResolver.getFullyQualifiedName(any())).thenReturn("::Foo");

    CppTypeInfo mapped = typeMapper.mapType(francaTypeRef);

    assertEquals("::Foo", mapped.name);
  }

  @Test
  public void mapTypeDefPredefine() {
    when(francaTypeRef.getDerived()).thenReturn(francaTypeDef);
    when(francaTypeDef.getActualType()).thenReturn(francaTypeRef2);
    when(francaTypeRef2.getPredefined()).thenReturn(FBasicTypeId.BOOLEAN);

    CppTypeInfo mapped = typeMapper.mapType(francaTypeRef);

    assertEquals(CType.BOOL.name, mapped.name);
  }

  @Test
  public void mapStructType() {
    when(francaTypeRef.getDerived()).thenReturn(francaStructType);
    when(cppNameResolver.getFullyQualifiedName(any())).thenReturn("::Foo");

    CppTypeInfo resultType = typeMapper.mapType(francaTypeRef);

    assertEquals("::Foo", resultType.name);
    assertTrue(resultType.includes.contains(CPP_INCLUDE));
    assertTrue(resultType.includes.contains(CBRIDGE_INCLUDE));
  }

  @Test
  public void mapPredefinedType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.BOOLEAN);

    CppTypeInfo mapped = typeMapper.mapType(francaTypeRef);
    assertEquals(CType.BOOL.name, mapped.name);
  }

  @Test
  public void mapEnumerationType() {
    when(francaTypeRef.getDerived()).thenReturn(francaEnum);
    when(cppNameResolver.getFullyQualifiedName(any())).thenReturn("::Foo");

    CppTypeInfo resultType = typeMapper.mapType(francaTypeRef);

    assertEquals("::Foo", resultType.name);
    assertTrue(resultType.includes.contains(CPP_INCLUDE));
    assertTrue(resultType.includes.contains(CBRIDGE_INCLUDE));
  }

  @Test
  public void mapArrayInlineType() {
    when(francaTypeRef.getPredefined()).thenReturn(FBasicTypeId.UINT64);
    when(FrancaTypeHelper.isImplicitArray(any())).thenReturn(true);

    CppTypeInfo actualType = typeMapper.mapType(francaTypeRef);

    assertEquals(CppTypeInfo.TypeCategory.ARRAY, actualType.typeCategory);
    assertTrue(actualType instanceof CppArrayTypeInfo);

    CppArrayTypeInfo actualArrayType = (CppArrayTypeInfo) actualType;
    assertNotNull(actualArrayType.innerType);
    assertEquals(actualArrayType.innerType.functionReturnType.name, "uint64_t");
  }

  @Test
  public void mapMapTypeWithPrimitiveKeyType() {
    when(francaTypeRef.getDerived()).thenReturn(francaMap);
    when(francaMap.getKeyType().getDerived()).thenReturn(null);
    when(francaMap.getKeyType().getPredefined()).thenReturn(FBasicTypeId.STRING);
    when(francaMap.getValueType().getDerived()).thenReturn(francaStructType);
    when(cppNameResolver.getFullyQualifiedName(francaStructType)).thenReturn("::Foo");
    when(cppNameResolver.getFullyQualifiedName(francaMap)).thenReturn("::FooMap");

    CppTypeInfo result = typeMapper.mapType(francaTypeRef);

    assertEquals(CppTypeInfo.TypeCategory.MAP, result.typeCategory);
    assertTrue(result instanceof CppMapTypeInfo);

    CppMapTypeInfo resultMapType = (CppMapTypeInfo) result;
    assertEquals("::FooMap", resultMapType.name);
    assertEquals(BASE_REF_NAME, resultMapType.cType.name);
    assertEquals(BASE_REF_NAME, resultMapType.functionReturnType.name);
    assertEquals("std::unordered_map<std::string, ::Foo>", resultMapType.baseApi);
    assertEquals(2, resultMapType.includes.size());
    assertEquals(BASE_HANDLE_IMPL_FILE, resultMapType.includes.get(0).getFileName());
    assertEquals(CppLibraryIncludes.MAP, resultMapType.includes.get(1));
  }

  @Test
  public void mapMapTypeWithPrimitiveEnumType() {
    when(francaTypeRef.getDerived()).thenReturn(francaMap);
    when(francaMap.getKeyType().getDerived()).thenReturn(francaEnum);
    when(francaMap.getValueType().getDerived()).thenReturn(francaStructType);
    when(cppNameResolver.getFullyQualifiedName(francaEnum)).thenReturn("::Foo");
    when(cppNameResolver.getFullyQualifiedName(francaStructType)).thenReturn("::Bar");

    CppTypeInfo result = typeMapper.mapType(francaTypeRef);

    assertEquals(CppTypeInfo.TypeCategory.MAP, result.typeCategory);
    assertTrue(result instanceof CppMapTypeInfo);

    CppMapTypeInfo resultMapType = (CppMapTypeInfo) result;
    assertEquals("std::unordered_map<::Foo, ::Bar, ::FooHash>", resultMapType.baseApi);
    assertEquals(3, resultMapType.includes.size());
    assertEquals(CppLibraryIncludes.ENUM_HASH, resultMapType.includes.get(2));
  }
}

File: genium/src/test/java/com/here/genium/generator/cbridge/CBridgeTypeMapperTest.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cbridge

import com.here.genium.generator.cbridge.CBridgeNameRules.BASE_REF_NAME
import com.here.genium.generator.cpp.CppIncludeResolver
import com.here.genium.generator.cpp.CppLibraryIncludes
import com.here.genium.generator.cpp.CppNameResolver
import com.here.genium.model.cbridge.CBridgeIncludeResolver
import com.here.genium.model.cbridge.CType
import com.here.genium.model.common.Include
import com.here.genium.model.lime.LimeArray
import com.here.genium.model.lime.LimeBasicType.TypeId
import com.here.genium.model.lime.LimeBasicTypeRef
import com.here.genium.model.lime.LimeDirectTypeRef
import com.here.genium.model.lime.LimeEnumeration
import com.here.genium.model.lime.LimeMap
import com.here.genium.model.lime.LimeNamedElement
import com.here.genium.model.lime.LimePath
import com.here.genium.model.lime.LimePath.Companion.EMPTY_PATH
import com.here.genium.model.lime.LimeTypeDef
import com.here.genium.test.AssertHelpers.assertContains
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class CBridgeTypeMapperTest {
    @MockK private lateinit var cppIncludeResolver: CppIncludeResolver
    @MockK private lateinit var cppNameResolver: CppNameResolver
    @MockK private lateinit var includeResolver: CBridgeIncludeResolver

    private lateinit var typeMapper: CBridgeTypeMapper

    @Before
    fun setUp() {
        MockKAnnotations.init(this, relaxed = true)

        typeMapper =
            CBridgeTypeMapper(cppIncludeResolver, cppNameResolver, includeResolver, "FooHash")
    }

    @Test
    fun mapTypeTypeDef() {
        val limeElement = LimeTypeDef(EMPTY_PATH, typeRef = LimeBasicTypeRef.FLOAT)

        val result = typeMapper.mapType(limeElement)

        assertEquals(CType.FLOAT, result.cType)
    }

    @Test
    fun mapTypeTypeDefToMapType() {
        val limeMap = LimeMap(LimeBasicTypeRef(TypeId.STRING), LimeBasicTypeRef.DOUBLE)
        val limeElement = LimeTypeDef(EMPTY_PATH, typeRef = LimeDirectTypeRef(limeMap))
        every { cppNameResolver.getFullyQualifiedName(limeElement) } returns "bar.Baz"

        val result = typeMapper.mapType(limeElement)

        assertEquals("bar.Baz", result.name)
        assertEquals(BASE_REF_NAME, result.cType.name)
        assertEquals(BASE_REF_NAME, result.functionReturnType.name)
        assertEquals(2, result.includes.size)
        assertEquals(CBridgeNameRules.BASE_HANDLE_IMPL_FILE, result.includes.first().fileName)
        assertEquals(CppLibraryIncludes.MAP, result.includes.last())
        assertEquals("std::string", (result as CppMapTypeInfo).keyType.name)
        assertEquals("double", result.valueType.name)
        assertNull(result.enumHashType)
    }

    @Test
    fun mapTypeTypeDefToMapTypeWithEnum() {
        val limeEnum = LimeEnumeration(LimePath(emptyList(), listOf("bar")))
        val limeMap = LimeMap(LimeDirectTypeRef(limeEnum), LimeBasicTypeRef.DOUBLE)
        val limeElement = LimeTypeDef(EMPTY_PATH, typeRef = LimeDirectTypeRef(limeMap))
        every { cppNameResolver.getFullyQualifiedName(limeEnum) } returns "Baz"
        every { cppNameResolver.getFullyQualifiedName(limeElement) } returns ""

        val result = typeMapper.mapType(limeElement)

        assertEquals(3, result.includes.size)
        assertEquals(CppLibraryIncludes.ENUM_HASH, result.includes.last())
        assertEquals("Baz", (result as CppMapTypeInfo).keyType.name)
        assertEquals("double", result.valueType.name)
        assertEquals("FooHash", result.enumHashType)
    }

    @Test
    fun mapTypeArray() {
        val limeElement = LimeArray(LimeBasicTypeRef.FLOAT)

        val result = typeMapper.mapType(limeElement)

        assertEquals("std::vector<float>", result.name)
        assertEquals(BASE_REF_NAME, result.cType.name)
        assertEquals(BASE_REF_NAME, result.functionReturnType.name)
        assertEquals(2, result.includes.size)
        assertEquals(CBridgeNameRules.BASE_HANDLE_IMPL_FILE, result.includes.first().fileName)
        assertEquals(CppLibraryIncludes.VECTOR, result.includes.last())
        assertEquals(CType.FLOAT, (result as CppArrayTypeInfo).innerType.cType)
    }

    @Test
    fun createCustomTypeInfo() {
        val limeElement = object : LimeNamedElement(EMPTY_PATH) {}
        val barInclude = Include.createInternalInclude("Bar")
        val bazInclude = Include.createInternalInclude("Baz")
        every { cppNameResolver.getFullyQualifiedName(limeElement) } returns "Foo"
        every { includeResolver.resolveInclude(limeElement) } returns barInclude
        every { cppIncludeResolver.resolveIncludes(limeElement) } returns listOf(bazInclude)

        val result = typeMapper.createCustomTypeInfo(limeElement, CppTypeInfo.TypeCategory.STRUCT)

        assertEquals("Foo", result.name)
        assertEquals(BASE_REF_NAME, result.cType.name)
        assertEquals(barInclude, result.cType.includes.first())
        assertEquals(BASE_REF_NAME, result.functionReturnType.name)
        assertEquals(barInclude, result.functionReturnType.includes.first())
        assertEquals(CppTypeInfo.TypeCategory.STRUCT, result.typeCategory)
        assertContains(barInclude, result.includes)
        assertContains(bazInclude, result.includes)
        assertContains(CBridgeTypeMapper.BASE_HANDLE_IMPL_INCLUDE, result.includes)
        assertContains(CppLibraryIncludes.OPTIONAL, result.includes)
        assertContains(CppLibraryIncludes.MEMORY, result.includes)
        assertContains(CppLibraryIncludes.NEW, result.includes)
    }

    @Test
    fun createEnumTypeInfo() {
        val limeElement = LimeEnumeration(LimePath(emptyList(), listOf("foobar", "barbaz")))
        val barInclude = Include.createInternalInclude("Bar")
        val bazInclude = Include.createInternalInclude("Baz")
        every { cppNameResolver.getFullyQualifiedName(limeElement) } returns "nonsense"
        every { includeResolver.resolveInclude(limeElement) } returns barInclude
        every { cppIncludeResolver.resolveIncludes(limeElement) } returns listOf(bazInclude)

        val result = typeMapper.createEnumTypeInfo(limeElement)

        assertEquals("nonsense", result.name)
        assertEquals("Foobar_Barbaz", result.cType.name)
        assertEquals(barInclude, result.cType.includes.first())
        assertEquals("Foobar_Barbaz", result.functionReturnType.name)
        assertEquals(barInclude, result.functionReturnType.includes.first())
        assertEquals(CppTypeInfo.TypeCategory.ENUM, result.typeCategory)
        assertContains(barInclude, result.includes)
        assertContains(bazInclude, result.includes)
    }
}
